feat: add db-connector powered virtual cluster instances restore procedure (#1955)
loft-sh/vcluster-docs: platform/administer/connector/database.mdx

  user: root
  type: mysql # This can be mysql or postgres. If left blank, the mysql database type is assumed.
```

## Recreate / restore virtual clusters

To accommodate infrastructure updates or disaster recovery needs, follow this process to restore virtual clusters using a **database connector** powered backing store.

<!-- vale Vale.Terms = NO -->
<Flow>

>   <Step>Annotate the target VirtualClusterInstance with `vcluster.com/database-reclaim-policy: retain` so the database is not deleted when the virtual cluster is deleted.</Step>

Reviewer: ```suggestion
  <Step>Annotate the target VirtualClusterInstance with `vcluster.com/database-reclaim-policy: retain` so deleting the virtual cluster does not also delete the database.</Step>
```

  <Step>Record the value of the `vcluster.com/database-uid` annotation and save the `spec` from the VirtualClusterInstance.</Step>
  <Step>Delete the virtual cluster.</Step>
  <Step>Create the virtual cluster with the same `vcluster.com/database-uid` annotation and the saved `spec`.</Step>
</Flow>
<!-- vale Vale.Terms = YES -->

:::important Scope
This procedure only restores the backing store connection. Depending on your configuration, resources such as persistent volumes or other external dependencies might not be available for the recreated virtual cluster.
:::

:::important RBAC permissions
Creating VirtualClusterInstance resources with the `vcluster.com/database-uid` annotation set requires elevated permissions.

```yaml title="Sample ClusterRole"
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: vcluster-database-admin
rules:
   - apiGroups: [ "management.loft.sh" ]
     resources: [ "virtualclusterinstances/externaldatabase" ]
     verbs: [ "bind" ]
```
:::

### Sample bulk migration script

:::important
This script assumes that every matched virtual cluster uses a database connector.

> If a matched virtual cluster does not use a database connector, its data will not be preserved.

Reviewer: ```suggestion
If a matched virtual cluster does not use a database connector, its data is not preserved.
```

:::

```sh title="Sample control plane host bulk migration script"

> # requires kubectl, jq and access to the vCluster Platform api

Reviewer: ```suggestion
# requires kubectl, jq, and access to the vCluster Platform api
```
